File: morse/comp/morse_shoot.f
      PROGRAM MORSE_SHOOT
C=======================================================================
C     ESPECTRO DEL POTENCIAL DE MORSE VIA SHOOTING METHOD (RK4)
C     VERSION ESTANDARIZADA PARA BENCHMARK HPC
C=======================================================================
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      
      CHARACTER*10 ARG_STR
      INTEGER NSTEPS, LEVEL, ITER
      DOUBLE PRECISION T1, T2, D_POT, ALPHA, XMIN, XMAX, DX
      DOUBLE PRECISION E_CUR, E_MIN, E_MAX, DE, F_PREV, F_CUR
      DOUBLE PRECISION EA, EB, FA, FMID, EMID
      DOUBLE PRECISION EN_LEVELS(5)

C     1. LECTURA DEL PARÁMETRO N (PASOS DE INTEGRACIÓN)
      CALL GET_COMMAND_ARGUMENT(1, ARG_STR)
      READ(ARG_STR, *) NSTEPS

C     PARAMETROS FISICOS (Mantenidos del original)
      D_POT = 10.0D0
      ALPHA = 0.5D0
      XMIN  = -2.0D0
      XMAX  = 25.0D0 
      DX = (XMAX - XMIN) / DBLE(NSTEPS)
      
      E_MIN = 0.0D0
      E_MAX = 10.0D0 
      DE    = 0.001D0 

C     --- INICIO DE MEDICIÓN ---
      CALL CPU_TIME(T1)

      E_CUR = E_MIN
      F_PREV = FUN_WAVE(E_CUR, XMIN, XMAX, DX, NSTEPS, D_POT, ALPHA)
      LEVEL = 0
      
C     2. BÚSQUEDA DE NIVELES (DISPARO + BISECCIÓN)
      DO WHILE (E_CUR .LT. E_MAX .AND. LEVEL .LT. 5)
         E_CUR = E_CUR + DE
         F_CUR = FUN_WAVE(E_CUR, XMIN, XMAX, DX, NSTEPS, D_POT, ALPHA)
         
         IF (F_PREV * F_CUR .LT. 0.D0) THEN
            EA = E_CUR - DE
            EB = E_CUR
            FA = F_PREV
            
            DO 10 ITER=1, 100 
               EMID = 0.5D0 * (EA + EB)
               FMID = FUN_WAVE(EMID,XMIN,XMAX,DX,NSTEPS,D_POT,ALPHA)
               IF (FA * FMID .LT. 0.D0) THEN
                  EB = EMID
               ELSE
                  EA = EMID
                  FA = FMID
               END IF
 10         CONTINUE
            
            LEVEL = LEVEL + 1
            IF (LEVEL .LE. 5) EN_LEVELS(LEVEL) = EMID
         END IF
         F_PREV = F_CUR
      END DO

C     --- FIN DE MEDICIÓN ---
      CALL CPU_TIME(T2)

C     3. SALIDA ESTANDARIZADA (N, E0, E1, E2, E3, E4, TIEMPO)
      WRITE(*,100) NSTEPS, (EN_LEVELS(I), I=1,4), (T2-T1)
100   FORMAT(I8, 5(1X, F20.12))

      END

C=======================================================================
      DOUBLE PRECISION FUNCTION FUN_WAVE(EN, X0, XF, H, N, D, ALP)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      
      PSI = 0.0D0     
      PHI = 1.0D-6    
      X = X0
      
      DO 20 I=1, N
         RK1_PSI = PHI
         RK1_PHI = 2.0D0 * (V_MORSE(X, D, ALP) - EN) * PSI
         
         XM2 = X + 0.5D0*H
         V_M = 2.0D0 * (V_MORSE(XM2, D, ALP) - EN)
         RK2_PSI = PHI + 0.5D0*H*RK1_PHI
         RK2_PHI = V_M * (PSI + 0.5D0*H*RK1_PSI)
         
         RK3_PSI = PHI + 0.5D0*H*RK2_PHI
         RK3_PHI = V_M * (PSI + 0.5D0*H*RK2_PSI)
         
         XF1 = X + H
         V_E = 2.0D0 * (V_MORSE(XF1, D, ALP) - EN)
         RK4_PSI = PHI + H*RK3_PHI
         RK4_PHI = V_E * (PSI + H*RK3_PSI)
         
         PSI = PSI + (H/6.0D0)*(RK1_PSI + 2.0D0*RK2_PSI + 
     &         2.0D0*RK3_PSI + RK4_PSI)
         PHI = PHI + (H/6.0D0)*(RK1_PHI + 2.0D0*RK2_PHI + 
     &         2.0D0*RK3_PHI + RK4_PHI)
         X = X + H
         
C        Normalización preventiva para evitar desbordamiento (Overflow)
         IF (DABS(PSI) .GT. 1.0D15) THEN
            PHI = PHI / 1.0D15
            PSI = PSI / 1.0D15
         END IF
 20   CONTINUE
      
      FUN_WAVE = PSI
      RETURN
      END

C=======================================================================
      DOUBLE PRECISION FUNCTION V_MORSE(X, D, ALP)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      V_MORSE = D * (1.0D0 - DEXP(-ALP * X))**2
      RETURN
      END

File: morse/Optimizacion_serial1/Obtimizacion _serial /morse_shoot.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

/* Función de Potencial */
static inline double v_morse(double x, double D, double alpha) {
    double tmp = 1.0 - exp(-alpha * x);
    return D * tmp * tmp;  // más eficiente que pow
}

/* Integrador Runge-Kutta de 4to Orden */
double fun_wave(double en, double x0, double xf, double h, int n, double D, double alpha) {
    double psi = 0.0;
    double phi = 1e-6;
    double x = x0;

    for (int i = 0; i < n; i++) {
        double rk1_psi = phi;
        double rk1_phi = 2.0 * (v_morse(x, D, alpha) - en) * psi;

        double xm2 = x + 0.5 * h;
        double v_m = 2.0 * (v_morse(xm2, D, alpha) - en);
        double rk2_psi = phi + 0.5 * h * rk1_phi;
        double rk2_phi = v_m * (psi + 0.5 * h * rk1_psi);

        double rk3_psi = phi + 0.5 * h * rk2_phi;
        double rk3_phi = v_m * (psi + 0.5 * h * rk2_psi);

        double xf1 = x + h;
        double v_e = 2.0 * (v_morse(xf1, D, alpha) - en);
        double rk4_psi = phi + h * rk3_phi;
        double rk4_phi = v_e * (psi + h * rk3_psi);

        psi += (h / 6.0) * (rk1_psi + 2.0 * rk2_psi + 2.0 * rk3_psi + rk4_psi);
        phi += (h / 6.0) * (rk1_phi + 2.0 * rk2_phi + 2.0 * rk3_phi + rk4_phi);
        x += h;

        if (fabs(psi) > 1e15) {
            psi /= 1e15;
            phi /= 1e15;
        }
    }
    return psi;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Uso: %s <NSTEPS>\n", argv[0]);
        return 1;
    }

    int nsteps = atoi(argv[1]);
    double d_pot = 10.0, alpha = 0.5;
    double xmin = -2.0, xmax = 25.0;
    double dx = (xmax - xmin) / (double)nsteps;

    double e_min = 0.0, e_max = 10.0, de = 0.001;
    double en_levels[5] = {0};
    int level = 0;

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);

    double e_cur = e_min;
    double f_prev = fun_wave(e_cur, xmin, xmax, dx, nsteps, d_pot, alpha);

    while (e_cur < e_max && level < 5) {
        e_cur += de;
        double f_cur = fun_wave(e_cur, xmin, xmax, dx, nsteps, d_pot, alpha);

        if (f_prev * f_cur < 0.0) {
            double ea = e_cur - de;
            double eb = e_cur;
            double fa = f_prev;
            double emid;

            for (int iter = 0; iter < 100; iter++) {
                emid = 0.5 * (ea + eb);
                double fmid = fun_wave(emid, xmin, xmax, dx, nsteps, d_pot, alpha);
                if (fa * fmid < 0.0) {
                    eb = emid;
                } else {
                    ea = emid;
                    fa = fmid;
                }
            }
            if (level < 5) en_levels[level] = emid;
            level++;
        }
        f_prev = f_cur;
    }

    clock_gettime(CLOCK_MONOTONIC, &end);
    double time_used = (end.tv_sec - start.tv_sec) +
                       (end.tv_nsec - start.tv_nsec) / 1e9;

    printf("%8d %20.12f %20.12f %20.12f %20.12f %20.12f\n", 
           nsteps, en_levels[0], en_levels[1],
           en_levels[2], en_levels[3], time_used);

    return 0;
}